pkg/apiserver/api_installer.go

```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
 	return ws
 }
 
-func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storage, ws *restful.WebService, proxyHandler http.Handler) (*unversioned.APIResource, error) {
-	admit := a.group.Admit
-	context := a.group.Context
-
-	optionsExternalVersion := a.group.GroupVersion
-	if a.group.OptionsExternalVersion != nil {
-		optionsExternalVersion = *a.group.OptionsExternalVersion
-	}
-
-	var resource, subresource string
-	switch parts := strings.Split(path, "/"); len(parts) {
-	case 2:
-		resource, subresource = parts[0], parts[1]
-	case 1:
-		resource = parts[0]
-	default:
-		// TODO: support deeper paths
-		return nil, fmt.Errorf("api_installer allows only one or two segment paths (resource or resource/subresource)")
+// validFQKind validates and returns the group version kind registered for the given storage
+// object. If the storage object is a subresource and has an override specified, the object's
+// group version is validated against the group version in the override for that subresource.
+func (a *APIInstaller) validFQKind(path string, storage rest.Storage) (unversioned.GroupVersionKind, error) {
+	fqKindToRegister := unversioned.GroupVersionKind{}
+	allowedGroup := a.group.GroupVersion.Group
+	if override, ok := a.group.SubresourceOverrides[path]; ok {
+		allowedGroup = override.GroupVersion.Group
 	}
-	hasSubresource := len(subresource) > 0
 
 	object := storage.New()
 	fqKinds, err := a.group.Typer.ObjectKinds(object)
 	if err != nil {
-		return nil, err
+		return fqKindToRegister, err
 	}
+
 	// a given go type can have multiple potential fully qualified kinds.  Find the one that corresponds with the group
 	// we're trying to register here
-	fqKindToRegister := unversioned.GroupVersionKind{}
 	for _, fqKind := range fqKinds {
-		if fqKind.Group == a.group.GroupVersion.Group {
-			fqKindToRegister = fqKind
+		// DEBUG: if fqKind.Group == a.group.GroupVersion.Group
+		if fqKind.Group == allowedGroup {
+			fqKindToRegister.Group = fqKind.Group
+			fqKindToRegister.Version = a.group.GroupVersion.Version
+			fqKindToRegister.Kind = fqKind.Kind
 			break
 		}
 
 		// TODO This keeps it doing what it was doing before, but it doesn't feel right.
 		if fqKind.Group == extensions.GroupName && fqKind.Kind == "ThirdPartyResourceData" {
```

Comment:
Let's deal with ThirdPartyResource in another PR. The currently implemented behavior isn't what we want.